src/modules/function/declaration.rs

```diff
@@ -205,21 +202,23 @@ impl SyntaxModule<ParserMetadata> for FunctionDeclaration {
                     };
                 }
                 let mut returns_tok = None;
+                let mut declared_failable = false;
                 // Optionally parse the return type
                 match token(meta, ":") {
                     Ok(_) => {
                         returns_tok = meta.get_current_token();
-                        self.returns = parse_type(meta)?
+                        self.returns = parse_type(meta)?;
+                        if token(meta, "?").is_ok() {
+                            declared_failable = true;
+                        }
                     },
                     Err(_) => self.returns = Type::Generic
                 }
                 // Parse the body
                 token(meta, "{")?;
-                let (index_begin, index_end, is_failable) = skip_function_body(meta);
-                if is_failable && !matches!(self.returns, Type::Failable(_) | Type::Generic) {
-                    return error!(meta, returns_tok, "Failable functions must return a Failable type");
-                } else if !is_failable && matches!(self.returns, Type::Failable(_)) {
-                    return error!(meta, returns_tok, "Non-failable functions cannot return a Failable type");
+                let (index_begin, index_end, is_failable) = skip_function_body(meta, declared_failable, &returns_tok)?;
+                if (!is_failable) && declared_failable {
+                    return error!(meta, returns_tok, "Infallible functions must not have a '?' after the type name");
```

Comment:
Consider changing to "Functions that cannot fail must not have a '?' after the type name" for consistency with the other error message.